src/hotspot/share/memory/arena.cpp

```diff
@@ -335,8 +335,8 @@ void *Arena::Arealloc(void* old_ptr, size_t old_size, size_t new_size, AllocFail
   size_t corrected_new_size = ARENA_ALIGN(new_size);
 
   // See if we can resize in-place
-  if( (c_old+old_size == _hwm) &&       // Adjusting recent thing
-      (c_old+corrected_new_size <= _max) ) {      // Still fits where it sits
+  if( (c_old+old_size == _hwm) &&                      // Adjusting recent thing
+      ((size_t)(_max-c_old) >= corrected_new_size) ) { // Still fits where it sits, safe from overflow
```

Comment:
This change is correct, but it hides an important finding behind a reshuffling of parameters that someone else may innocently reshape later. And in the process, makes the code less readable. Can we use something like saturated_add()?

I would also add an explicit assert for a reasonable max size. Arena allocations should be small. Nobody should hand in sizes larger than a few MB, so asserting for size >= 2^31 (2g) would make sense. Anything as large as that is almost certainly an error we should trap on.

Reply:
I would support this.

May be possible to just use `saturated_add()`. Would be limited to `max_jint`, but with an assert for sizes above 2^31, that shouldn't be an issue. Will look into it.

Reply:
Please, not `max_jint`.  This limit isn't tied to Java integer semantics.  `MAX_INT` might be appropriate.
(Yes, I know they likely have the same value.)

Reply:
@tstuefe I've now added asserts and changed this check to use saturated adds.

The already existing `saturated_add()` only works on `int`s and `uint`s, so I made a version working on offsetting pointers. We still have to check for overflow (if the pointer addition was saturated), but the code becomes more readable.